### PR TITLE
Implement generic "KVS player" downloader

### PR DESCRIPTION
**Pull request by rigstot**
KVS Player is a generic youtube clone/tube site software. A SFW demo
instance is hosted on https://www.kvs-demo.com/.

related issue on youtube-dl: https://github.com/ytdl-org/youtube-dl/issues/2077

This replaces the Thisvid-specific extractor I wrote for youtube-dlc.

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This replaces the earlier extractor I wrote for Thisvid(.)com, and makes it a generic downloader for all tube sites based on KVS Player.

### CODE REVIEW

**Comment by rigstot** [comment]
Updated to allow non-obfuscated URLs and formats with missing ID.

File: yt_dlp/extractor/generic.py [hunk 1]
```diff
@@ -2238,6 +2238,83 @@ class GenericIE(InfoExtractor):
                 'title': '#WEAREFMI – PT.2 – 2021 – MsMotorTV',
             },
             'playlist_count': 1,
+        }, {
+            # KVS Player (1/7)
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:2242]: Remove `(1/7)` and similar

File: yt_dlp/extractor/generic.py [hunk 2]
```diff
@@ -2238,6 +2238,83 @@ class GenericIE(InfoExtractor):
...
+            'url': 'https://www.kvs-demo.com/videos/105/kelis-4th-of-july/',
+            # md5 changes every time
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:2244]: then use `skip_download`

File: yt_dlp/extractor/generic.py [hunk 3]
```diff
@@ -3478,6 +3593,55 @@ def filter_video(urls):
                 )
                 .*?
                 ['"]?file['"]?\s*:\s*["\'](.*?)["\']''', webpage))
+        if not found:
+            # Look for generic KVS player (must be before 'Broaden ...' sections)
+            found = re.search(r'<script [^>]*?src="https://.+?/kt_player\.js\?v=(?P<ver>(?P<maj_ver>\d+)(\.\d+)+)".*?>', webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3598]: > (must be before 'Broaden ...' sections)

?

File: yt_dlp/extractor/generic.py [hunk 4]
```diff
@@ -3478,6 +3593,55 @@ def filter_video(urls):
...
+            if found:
+                if found.group('maj_ver') not in ['4', '5']:
+                    self.report_warning("Untested major version (" + found.group('ver') + ") in player engine--Download may fail.")
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3601]: use `%`, not `+`
and `'`, not `"`

File: yt_dlp/extractor/generic.py [hunk 5]
```diff
@@ -3478,6 +3593,55 @@ def filter_video(urls):
...
+                flashvars = re.search(r'(?ms)<script.*?>.*?var\s+flashvars\s*=\s*(\{.*?\});.*?</script>', webpage)
+                flashvars = self._parse_json(flashvars.group(1), video_id, transform_source=js_to_json)
+
+                # extract the part after the last / as the display_id from the
+                # canonical URL. handles cases where rel comes before and after
+                # the href attribute
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3607]:
```suggestion
                # extract the part after the last / as the display_id from the canonical URL
```

File: yt_dlp/extractor/generic.py [hunk 6]
```diff
@@ -3478,6 +3593,55 @@ def filter_video(urls):
...
+                thumbnail = flashvars['preview_url']
+                if thumbnail.startswith("//"):
+                    protocol, _, _ = url.partition('/')
+                    thumbnail = protocol + thumbnail
+
+                def extract_height(url, video_id):
+                    found = re.search(r'%s_(\d+)p\.mp4/?$' % video_id, url)
+                    if found:
+                        return int(found.group(1))
+                    return float('inf')  # Hack: if no height-suffix, it is the largest
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3624]: don't do this. Instead set quality=1 to prioritize it over other formats

File: yt_dlp/extractor/generic.py [hunk 7]
```diff
@@ -3478,6 +3593,55 @@ def filter_video(urls):
...
+
+                formats = [
+                    {
+                        'url': self._kvs_getrealurl(flashvars[key], flashvars['license_code']),
+                        'format_id': flashvars.get(key + '_text', key),
+                        'height': extract_height(flashvars[key], flashvars['video_id']),
+                        'ext': 'mp4',
+                    }
+                    for key in ("video_url", "video_alt_url", "video_alt_url2")
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3633]: Use `'` everywhere

File: yt_dlp/extractor/generic.py [hunk 8]
```diff
@@ -3478,6 +3593,55 @@ def filter_video(urls):
...
+                    if key in flashvars and "/get_file/" in flashvars[key]
+                ]
+                formats = sorted(formats, key=lambda f: f['height'])
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3636]: no. Use `self._sort_formats(formats)`

File: yt_dlp/extractor/thisvid.py [hunk 9]
```diff
@@ -1,97 +0,0 @@
...
-        }
-    }, {
-        'url': 'https://thisvid.com/embed/2400174/',
-        'md5': '3397979512c682f6b85b3b04989df224',
-        'info_dict': {
-            'id': '2400174',
-            'ext': 'mp4',
-            'title': 'French Boy Pantsed',
-            'thumbnail': 'https://media.thisvid.com/contents/videos_screenshots/2400000/2400174/preview.mp4.jpg',
-            'age_limit': 18,
```
Comment (pukkandan) [inline yt_dlp/extractor/thisvid.py:28]: The `age_limit` information is lost when using generic. Make this a thin wrapper instead of deleting

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/generic.py (hunk 3 above)
Comment (rigstot) [reply yt_dlp/extractor/generic.py:3602]: There is a block called `# Broaden the search a little bit`. The KVS code must be placed above it.

File: yt_dlp/extractor/thisvid.py (hunk 9 above)
Comment (rigstot) [reply yt_dlp/extractor/thisvid.py:28]: `age_limit` was never really calculated, it was just hard-coded to 18+ every time. I don't think a `if url.startswith('https://thisvid.com')` is really useful.

**Comment by rigstot** [comment]
Thank you for the review, @pukkandan! I have fixed most of your review notes, but I would ask you to get back to me about the remaining two.

**Comment by pukkandan** [comment]
In the future, please avoid force pushing. It make it much harder for me to track the changes

**Review by pukkandan — APPROVED** [review]

**Comment by rigstot** [comment]
Thanks for merging and sorry that I missed some quotes.

> In the future, please avoid force pushing. It make it much harder for me to track the changes

Noted for the future. In case you were not aware of this: you can view changes after a force-push by clicking here:

    ￼ rigstot force-pushed the rigstot:master branch from fd34a0e to a01aed0 18 hours ago
              ^^^^^^^^^^^^

**Comment by pukkandan** [comment]
oh, I never knew that part was clickable. I always used to manually type in the url like https://github.com/yt-dlp/yt-dlp/compare/hash1..hash2

**Pull request merged**